### PR TITLE
Where Jón provides a new rememberMe cookie serializer

### PR DESCRIPTION
**Pull request by jontg**
See [this Shiro ticket](https://issues.apache.org/jira/browse/SHIRO-226) [and this gist](https://gist.github.com/duskote/0f8cce8d1b0d1cc56844).  The rememberMe cookie by default uses java serialization leading to pretty big cookie sizes.  That's clearly no good, so we fix that by switching to a gzip stream as suggested in the comments.

### CODE REVIEW

File: soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+                // Again, limited to 32768 principals.
+                out.writeShort(principals.size());
+
+                for (Object principal : principals) {
+                    out.writeObject(principal);
+                }
+            }
+            gout.finish();
+        } catch (IOException e) {
```
Comment (relateiq-ec2) [inline soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * See https://gist.github.com/duskote/0f8cce8d1b0d1cc56844 ; the rememberMe cookie by default uses java serialization
+ * leading to pretty big cookie sizes.  That's clearly no good, so we fix that by switching to a gzip stream.
+ */
+public class SimplePrincipalSerializer implements Serializer<PrincipalCollection> {
+    /**
+     * Magic number to signal that this is a SimplePrincipalSerializer file so that we don't try to decode something crap.
+     */
+    private static final int MAGIC = 0x0BADBEEF;
+
+    public byte[] serialize(PrincipalCollection pc) throws SerializationException {
```
Comment (relateiq-ec2) [inline soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java]: MINOR Remove the declaration of thrown exception 'org.apache.shiro.io.SerializationException' which is a runtime exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
MAJOR Add the "@Override" annotation above this method signature [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

File: soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                }
+            }
+            gout.finish();
+        } catch (IOException e) {
+            throw new SerializationException(e.getMessage());
+        }
+        return ba.toByteArray();
+    }
+
+    public PrincipalCollection deserialize(byte[] serialized) throws SerializationException {
```
Comment (relateiq-ec2) [inline soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java]: MINOR Remove the declaration of thrown exception 'org.apache.shiro.io.SerializationException' which is a runtime exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
MAJOR Add the "@Override" annotation above this method signature [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

**Comment by relateiq-ec2** [comment]
SonarQube analysis reported 5 issues:
* CRITICAL 1 critical
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by jontg** [comment]
Force-pushed a fix to address the sonarqube issues that were flagged.

The `rememberMe` functionality isn't yet used within the app, but if we decide to use it we should use a good implementation (rather than the built-in java serialization).

File: soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            GZIPOutputStream gout = new GZIPOutputStream(ba);
+            ObjectOutputStream out = new ObjectOutputStream(gout);
+
+            // Write the magic number which allows us to decode it later on
+            out.writeInt(MAGIC);
+
+            // Limited to 32768 realms. Should be enough for everybody.
+            out.writeShort(pc.getRealmNames().size());
+
+            for (String realm : pc.getRealmNames()) {
```
Comment (christydennison) [inline soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java]: I'm a fan of using forEach lambdas whenever possible, but it's super minor :)

File: soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+            // Check magic number
+            if (in.readInt() != MAGIC)
+                throw new SerializationException(
+                        "Not valid magic number while deserializing stored PrincipalCollection - possibly obsolete cookie.");
+
+            int numRealms = in.readShort();
+
+            // realms loop
+            for (int i = 0; i < numRealms; i++) {
```
Comment (christydennison) [inline soa-auth/src/main/java/com/relateiq/soa/auth/SimplePrincipalSerializer.java]: I'm a fan of IntStream but your call :)

**Comment by christydennison** [comment]
Looks awesome!! Couple of minor comments that you should feel free to fix or not. :)

**Comment by jontg** [comment]
I wanna keep this code as close to the [gist](https://gist.github.com/duskote/0f8cce8d1b0d1cc56844) as possible, so I'm going to merge as-is

**Pull request merged**